File: androidmvp/src/main/java/com/hsc/vince/androidmvp/base/BasePresenter.java
package com.hsc.vince.androidmvp.base;

import com.hsc.vince.androidmvp.net.ApiStore;
import com.hsc.vince.androidmvp.net.DecodeTransformer;
import com.hsc.vince.androidmvp.net.NetClient;

import org.reactivestreams.Subscriber;

import io.reactivex.Flowable;
import io.reactivex.FlowableTransformer;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;

/**
 * <p>作者：黄思程  2018/4/11 16:42
 * <p>邮箱：codinghuang@163.com
 * <p>作用：
 * <p>描述：基类Presenter
 * 观察者和被观察者绑定，结束绑定操作
 */
public class BasePresenter<V> {
    protected V mMvpView;
    protected ApiStore mApiStore;
    private CompositeDisposable mCompositeDisposable;

    /*** 创建并注入View
     * @param v view*/
    public BasePresenter(V v) {

        attachView(v);
    }

    /*** 绑定View
     * @param mvpView mvpView*/
    private void attachView(V mvpView) {
        this.mMvpView = mvpView;
        mApiStore = NetClient.newInstance().create(ApiStore.class);
    }

    /***解绑View*/
    public void detachView() {
        onUnsubscribe();
        this.mMvpView = null;
    }

    /*** RXjava取消注册，以避免内存泄露*/
    public void onUnsubscribe() {
        if (mCompositeDisposable != null && !mCompositeDisposable.isDisposed()) {
            mCompositeDisposable.dispose();
        }
    }

    /**
     * Rxjava添加注册，使用CompositeSubscription绑定被观察者和观察者，方便管理
     *
     * @param observable 被观察者，会使用到装饰者
     * @param subscriber 观察者
     */
    protected <T> void addSubscription(Flowable<T> observable, Subscriber<T> subscriber) {
        if (mCompositeDisposable == null) {
            mCompositeDisposable = new CompositeDisposable();
        }

        FlowableTransformer<T, T> transformer = new DecodeTransformer<>();
        Disposable subscribe = observable.compose(transformer).subscribe();
        mCompositeDisposable.add(subscribe);
    }
}

File: news/src/main/java/com/vince/news/base/BasePresenter.kt
package com.vince.news.base

import com.vince.news.constant.BoreTest
import com.vince.news.net.ApiStore
import com.vince.news.net.HttpManager
import rx.Observable
import rx.Subscriber
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import rx.subscriptions.CompositeSubscription

/**
 * <p>作者：黄思程  2017/11/16 16:44
 * <p>邮箱：huangsicheng@gouuse.cn
 * <p>作用：
 * <p>描述：Presenter基类
 */
open class BasePresenter<V>(v: V) {
    protected var mvpView: V? = null
    protected lateinit var apiStores: ApiStore
    private var mCompositeSubscription: CompositeSubscription? = null

    init {
        attachView(v)
    }

    /**
     * 绑定view，并创建apiStores
     * @param mvpView view*/
    private fun attachView(mvpView: V) {
        this.mvpView = mvpView
        apiStores = HttpManager.newInstance().create(ApiStore::class.java)
    }

    /*** 重置网络链接*/
    fun resetClient() {
        if (BoreTest.isUnitTest) {
            return
        }
        HttpManager.clear()
        apiStores = HttpManager.newInstance().create(ApiStore::class.java)
    }

    /*** 释放view，并解绑RxJava*/
    fun detachView() {
        unsubscribe()
        this.mvpView = null
    }

    /*** RxJava取消注册，以避免内存泄露*/
    fun unsubscribe() {
        if (mCompositeSubscription != null && mCompositeSubscription!!.hasSubscriptions()) {
            mCompositeSubscription?.unsubscribe()
        }
    }

    /**
     * RxJava添加注册，使用CompositeSubscription绑定被观察者和观察者，方便管理
     * @param observable 被观察者，会使用到装饰者
     * @param subscriber 观察者   */
    protected fun <T> addSubscription(observable: Observable<T>, subscriber: Subscriber<T>) {
        if (mCompositeSubscription == null) {
            mCompositeSubscription = CompositeSubscription()
        }
        val transformer: Observable.Transformer<T, T> = ObservableTransformer()
        val subscription = observable.compose(transformer).subscribe(subscriber)
        mCompositeSubscription?.add(subscription)
    }
}

class ObservableTransformer<T> : Observable.Transformer<T, T> {
    override fun call(observable: Observable<T>): Observable<T> {
        return if (BoreTest.isUnitTest) {
            observable
                    .subscribeOn(Schedulers.immediate())
                    .observeOn(Schedulers.immediate())
        } else {
            observable
                    .subscribeOn(Schedulers.io())
                    .unsubscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
        }
    }
}